logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,215 @@
+package org.logstash.secret.store.backend;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.logstash.secret.SecretIdentifier;
+import org.logstash.secret.store.SecretStore;
+import org.logstash.secret.store.SecretStoreException;
+import org.logstash.secret.store.SecretStoreUtil;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.KeyStore;
+import java.security.KeyStore.PasswordProtection;
+import java.security.KeyStore.ProtectionParameter;
+import java.security.KeyStoreException;
+import java.security.UnrecoverableKeyException;
+import java.util.*;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * <p>Java Key Store implementation for the {@link SecretStore}.</p>
+ * <p>Note this implementation should not be used for high volume or large datasets.</p>
+ * <p>This class is threadsafe.</p>
+ */
+public final class JavaKeyStore implements SecretStore {
+    public static final String LOGSTASH_MARKER = "logstash-key-store";
+    private static final Logger LOGGER = LogManager.getLogger(JavaKeyStore.class);
+    private static final String KEYSTORE_TYPE = "pkcs12";
+
+    private final char[] keyStorePass;
+    private final Path keyStorePath;
+    private final ProtectionParameter protectionParameter;
+    private final Lock readLock;
```

Comment:
Should this be a [FileLock](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileLock.html) (or in addition?) to avoid the keystore being mutated while we are using it? (Logstash reading, operator writing, for example)

Reply:
I don't think so, that would require a downtime to update the secret store.  The likely hood of collision is very low since these the only interaction with the key store should be at settings load and pipeline load time.  Even if there is a cross-JVM collision, i am not sure how the underlying implementation handles it, a quick scan of the code appears it would result in last action wins, but hard to say for certain.

Reply:
> that would require a downtime to update the secret store

My reading of the code is disagreeing with this claim.

readLock is used for listing and retrieving, for example. retrieveSecret does this:

* readLock.lock
* open the keystore file
* Load keystore
* do work
* unlock

This lock protects the entire file open/load/etc operation within the JVM. Extending this protection to the OS is a good idea (with FileLock)

I agree with you that there is an unlikely chance of a corrupted read-during-write given the store is only likely accessed during pipeline loads/reloads, etc.

My concern, for example, is someone copying in a new store *while* Logstash is starting up, possibly resulting in an InputStream that is corrupted because our reads are part from the old file, part from the new file.

We can add file locks in a future PR perhaps? I am not seeing any harm in adding safety.

Two safety ideas:
1) Write atomically (write to keystore.new + rename). This is what we do for PQ.
2) Use file locks.

Atomic writes are safer for system halts/faults where a write is interrupted. FileLock is sufficient for blocking writes while the file is being read.